website/src/documentation/programming-guide.md

```diff
@@ -1970,7 +1976,1078 @@ records.apply("WriteToText",
+  public long getCostCents();  // Returns the cost of the item.
+  public List<Transaction> getTransactions();  // Returns the transactions that paid for this purchase (returns a list, since the purchase might be spread out over multiple credit cards).
+  
+  @SchemaCreate
+  public Purchase(String userId, long itemId, ShippingAddress shippingAddress, long costCents, 
+                  List<Transaction> transactions) {
+      ...
+  }
+}
+
+@DefaultSchema(JavaBeanSchema.class)
+public class ShippingAddress {
+  public String getStreetAddress();
+  public String getCity();
+  @Nullable public String getState();
+  public String getCountry();
+  public String getPostCode();
+  
+  @SchemaCreate
+  public ShippingAddress(String streetAddress, String city, @Nullable String state, String country,
+                         String postCode) {
+     ...
+  }
+}
+
+@DefaultSchema(JavaBeanSchema.class)
+public class Transaction {
+  public String getBank();
+  public double getPurchaseAmount();
+ 
+  @SchemaCreate
+  public Transaction(String bank, double purchaseAmount) {
+     ...
+  }
+}
+```
+
+Using JavaBean classes as above is one way to map a schema to Java classes. However multiple Java classes might have
+the same schema, in which case the different Java types can often be used interchangeably. For example, the above
+`Transaction` class has the same schema as the following class:
+
+```java
+@DefaultSchema(JavaFieldSchema.class)
+public class TransactionPojo {
+  public String bank;
+  public double purchaseAmount;
+}
+```
+
+So if we had two `PCollection`s as follows
+
+```java
+PCollection<Transaction> transactionBeans = readTransactionsAsJavaBean();
+PCollection<TransactionPojos> transactionPojos = readTransactionsAsPojo();
+```
+
+Then these two `PCollection`s would have the same schema, even though their Java types would be different. This means
+for example the the following two code snippets are valid:
+
+```java
+transactionBeans.apply(ParDo.of(new DoFn<...>() {
+   @ProcessElement public void process(@Element TransactionPojo pojo) {
+      ...
+   }
+}));
+```
+
+and
+```java
+transactionPojos.apply(ParDo.of(new DoFn<...>() {
+   @ProcessElement public void process(@Element Transaction row) {
+    }
+}));
+```
+
+Even though the in both cases the `@Element` parameter differs from the the `PCollection`'s Java type, since the
+schemas are the same Beam will automatically make the conversion. The built-in `Convert` transform can also be used
+to translate between Java types of equivalent schemas, as detailed below.
+
+### 6.3. Schema definition {#schema-definition}
+The schema for a `PCollection` defines elements of that `PCollection` as an ordered list of named fields. Each field
+has a name, a type, and possibly a set of user options. The type of a field can be primitive or composite. The following
+are the primitive types currently supported by Beam:
```

Comment:
Are there any specific limitations for field names we want to document here? Many special characters (`$`, `-`, ...) make it impossible to map schemas to POJO and some other external systems, and others (`.`, `[`, `]`) make field accesses ambiguous and seems to be effectively disallowed in Java today.

Perhaps we _should_ allow all of these special characters in field names, and make that clear here. We would need to make sure our implementation provides ways to work around the issues - For example, allow escaping in field access descriptors.

Maybe this is worth a ML discussion..

Reply:
Looks like the consensus from [dev@](https://lists.apache.org/thread.html/re0451b85feb703d8305f7044d17da9cdc494efaa4b7c0fcf152514e4%40%3Cdev.beam.apache.org%3E) is field names should be allowed to contain any character. Can we document that somewhere?

Reply:
I would prefer to wait until we actually can support that (i.e. until our selection language supports escaping characters) before documenting that.